File: DataStructure/merge_sort2.py
def merge_sort(arr):
    if len(arr) < 2:
        return arr
    mid = len(arr)//2
    leftList = arr[:mid]
    rightList = arr[mid:]
    leftList = merge_sort(leftList)
    rightList = merge_sort(rightList)
    return merge(leftList, rightList)

def merge(l, h):
    result = []
    while len(l) > 0 or len(r) > 0:
        if len(l) > 0 and len(r) > 0:
            if l[0] <= r[0]:
                result.append(l[0])
                l = l[1:]
            else:
                result.append(r[0])
                r = r[1:]
        elif len(l) > 0:
            result.append(l[0])
            l = l[1:]
        elif len(r) > 0:
            result.append(r[0])
            r = r[1:]
    return result

File: DataStructure/MergeSort2.java
public class MergeSort{
    public static void mergeSort(int[] arr){
        sort(arr, 0, arr.length);
    }

    private static void sort(int[] arr, int l, int h){
        if(h - l < 2) return;

        int mid = (l + h) /2;
        sort(arr, 0, mid);
        sort(arr, mid, h);
        merge(arr, l, mid, h);
    }

    private static void merge(int[] arr, int l, int mid, int h){
        int[] temp = new int[h - l];
        int t = 0, low = l, high = mid;

        while(low < mid && high < h){
            if(arr[low] < arr[high]){
                temp[t++] = arr[low++];
            }else{
                temp[t++] = arr[high++];
            }
        }

        while(low < mid){
            temp[t++] = arr[low++];
        }

        while(high < h){
            temp[t++] = arr[high++];
        }

        for(int i = l; i < h; i++){
            arr[i] = temp[i - l];
        }
    }
}